hypershift-operator/controllers/nodepool/kubevirt/imagecacher.go

```diff
@@ -0,0 +1,247 @@
+	logger := ctrl.LoggerFrom(ctx)
+	for _, oldDV := range oldDVs {
+		if oldDV.DeletionTimestamp == nil {
+			logger.Info("deleting an old boot image cache DataVolume", "namespace", oldDV.Namespace, "DataVolume name", oldDV.Name)
+			err := cl.Delete(ctx, &oldDV)
+			if err != nil {
+				logger.Error(err, fmt.Sprintf("failed to delete an old DataVolume; namespace: %s, name: %s", oldDV.Namespace, oldDV.Name))
+			}
+		}
+	}
+}
+
+func (qi *cachedQCOWImage) createDVForCache(ctx context.Context, cl client.Client, nodePool *hyperv1.NodePool, infraId string) (*v1beta1.DataVolume, error) {
+	dv := qi.buildDVForCache(nodePool, infraId)
+
+	err := cl.Create(ctx, dv)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return nil, fmt.Errorf("failed to create a DataVolume for the boot image cache ; %w", err)
+	}
+
+	return dv, nil
+}
+
+func (qi *cachedQCOWImage) getDVSourceForVMTemplate() *v1beta1.DataVolumeSource {
+	return &v1beta1.DataVolumeSource{
+		PVC: &v1beta1.DataVolumeSourcePVC{
+			Namespace: qi.namespace,
+			Name:      qi.dvName,
+		},
+	}
+}
+
+func (qi *cachedQCOWImage) buildDVForCache(nodePool *hyperv1.NodePool, infraId string) *v1beta1.DataVolume {
+	var (
+		volumeSize   apiresource.Quantity
+		storageClass *string
+	)
+
+	kvPlatform := nodePool.Spec.Platform.Kubevirt
+
+	if kvPlatform.RootVolume != nil && kvPlatform.RootVolume.Persistent != nil {
+		if kvPlatform.RootVolume.Persistent.Size != nil {
+			volumeSize = *kvPlatform.RootVolume.Persistent.Size
+		}
+		storageClass = kvPlatform.RootVolume.Persistent.StorageClass
+	}
+
+	return &v1beta1.DataVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: bootImageNamePrefix,
+			Namespace:    qi.namespace,
+			Labels: map[string]string{
+				bootImageDVLabelRoleName: bootImageDVLabelRoleValue,
+				bootImageDVLabelInfraId:  infraId,
+			},
+			Annotations: map[string]string{
+				bootImageDVAnnotationHash:          qi.hash,
+				cdiImmediateBindingAnnotation:      "true",
+				cdiDeleteAfterCompletionAnnotation: "false",
+			},
+		},
+		Spec: v1beta1.DataVolumeSpec{
+			Source: &v1beta1.DataVolumeSource{
+				HTTP: &v1beta1.DataVolumeSourceHTTP{
+					URL: qi.name,
+				},
+			},
+			Preallocation: pointer.Bool(true),
+			Storage: &v1beta1.StorageSpec{
+				AccessModes: []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
+				Resources: corev1.ResourceRequirements{
+					Requests: map[corev1.ResourceName]apiresource.Quantity{
+						corev1.ResourceStorage: volumeSize,
+					},
+				},
+				StorageClassName: storageClass,
+			},
+		},
+	}
+}
+
+func getCacheDVSelector(infraId string) client.MatchingLabels {
+	return map[string]string{
+		bootImageDVLabelRoleName: bootImageDVLabelRoleValue,
+		bootImageDVLabelInfraId:  infraId,
+	}
+}
+
+func getCacheDVs(ctx context.Context, cl client.Client, infraId string) ([]v1beta1.DataVolume, error) {
+	dvs := &v1beta1.DataVolumeList{}
+
+	err := cl.List(ctx, dvs, getCacheDVSelector(infraId))
+
```

Comment:
This list should also be isolated to the namespace that the VMs and PVCs are in. 

add a `client.InNamespace(namespace)` to the options list.